lib/cpp-test/hawkes/inference/hawkes_em_gtest.cpp

```diff
@@ -0,0 +1,251 @@
+      EXPECT_DOUBLE_EQ(timefunc[uv].get_dt(), dt);
+    }
+  }
+}
+
+TEST_F(HawkesEMTest, kernel_time_func_data) {
+  em.init_kernel_time_func(kernels);
+  std::vector<TimeFunction>& timefunc = em.get_kernel_time_func();
+  ASSERT_FALSE(timefunc.empty());
+  for (ulong u = 0; u < n_nodes; ++u) {
+    for (ulong v = 0; v < n_nodes; ++v) {
+      ulong uv = u * n_nodes + v;
+      ArrayDouble data_uv = *(timefunc[uv].get_sampled_y());
+      EXPECT_EQ(data_uv.size(), kernel_size);
+      for (ulong k = 0; k < kernel_size; ++k) {
+        ulong vk = v * kernel_size + k;
+        // Test data values
+        EXPECT_DOUBLE_EQ(data_uv[k], kernels(u, vk))
+            << "Kernel[" << u << ", " << v << "]: "
+            << "Value of " << k << "-th sample data  gives a mismatch.";
+        // Test abscissa
+        EXPECT_DOUBLE_EQ(timefunc[uv].get_t0() + k * timefunc[uv].get_dt(),
+                         (*em.get_kernel_discretization())[k])
+            << "Kernel[" << u << ", " << v << "]: "
+            << "Value of " << k << "-th abscissa point  gives a mismatch.";
+      }
+    }
+  }
+}
+
+TEST_F(HawkesEMTest, kernel_time_func_data_with_explicit_abscissa) {
+  // Rounding errors in the division T[10] - T[0] can result in
+  // the size of `sampled_y` being larger than the size of `Y` by 1
+  em.set_kernel_discretization(kernel_discretization.as_sarray_ptr());
+  em.init_kernel_time_func(kernels);
+  std::vector<TimeFunction>& timefunc = em.get_kernel_time_func();
+  ASSERT_FALSE(timefunc.empty());
+  for (ulong u = 0; u < n_nodes; ++u) {
+    for (ulong v = 0; v < n_nodes; ++v) {
+      ulong uv = u * n_nodes + v;
+      ArrayDouble data_uv = *(timefunc[uv].get_sampled_y());
+      EXPECT_NEAR(data_uv.size(), kernel_size + 1, 1);
+      /*
+      std::cout << "Kernel[" << u << ", " << v << "]: " << std::endl;
+      for (ulong j = 0; j < data_uv.size(); j++) {
+        std::cout << "data_uv[" << j << "] = " << data_uv[j] << std::endl;
+      }
+      */
+      for (ulong k = 0; k < kernel_size; ++k) {
+        ulong vk = v * kernel_size + k;
+        // Test data values
+        EXPECT_DOUBLE_EQ(data_uv[k], kernels(u, vk))
+            << "Kernel[" << u << ", " << v << "]: " << std::endl
+            << "Value of " << k << "-th sample data  gives a mismatch." << std::endl
+            << "Corresponding kernel discretization point : "
+            << (*em.get_kernel_discretization())[k] << std::endl;
+      }
+    }
+  }
+}
+
+TEST_F(HawkesEMTest, kernel_time_func_values) {
+  em.init_kernel_time_func(kernels);
+  std::vector<TimeFunction>& timefunc = em.get_kernel_time_func();
+  ASSERT_FALSE(timefunc.empty());
+  for (ulong u = 0; u < n_nodes; ++u) {
+    for (ulong v = 0; v < n_nodes; ++v) {
+      ulong uv = u * n_nodes + v;
+      for (ulong k = 0; k < kernel_size; ++k) {
+        ulong vk = v * kernel_size + k;
+        double t = t0 + k * dt + .5 * dt;  // kernel_discretization[k] + .5 * dt;
```

Comment:
are these helpful in anyway?

Reply:
I think so. The evaluation of the Hawkes kernel is very consequential for the EM algorithm. Before these tests were in place, I could not really be sure about the correctness of the evaluation.

Reply:
ah, sure, that was meant to link the manual increment to the point in the array. I can take that out.

Reply:
Done here: e43af7e02cd75d01a935728a4fd81342b8f07d77